File: packages/rust/src/binary.rs
//! Binary Encoding
//!
//! Functions for encoding data to binary format for signing.

use base64::Engine;
use serde::Serialize;

use crate::canonicalize::canonicalize_bytes;
use crate::types::{Result, CONSTELLATION_PREFIX};

/// Convert data to bytes for signing
///
/// # Arguments
/// * `data` - Any serializable data
/// * `is_data_update` - Whether to encode as a DataUpdate (with Constellation prefix)
///
/// # Returns
/// UTF-8 bytes ready for hashing
///
/// # Example
/// ```
/// use constellation_sdk::binary::to_bytes;
/// use serde_json::json;
///
/// let data = json!({"id": "test"});
/// let bytes = to_bytes(&data, false).unwrap();
/// ```
pub fn to_bytes<T: Serialize>(data: &T, is_data_update: bool) -> Result<Vec<u8>> {
    let canonical_json = canonicalize_bytes(data)?;

    if is_data_update {
        // Add Constellation prefix for DataUpdate
        let base64_string = base64::engine::general_purpose::STANDARD.encode(&canonical_json);
        let wrapped_string = format!(
            "{}{}\n{}",
            CONSTELLATION_PREFIX,
            base64_string.len(),
            base64_string
        );
        Ok(wrapped_string.into_bytes())
    } else {
        Ok(canonical_json)
    }
}

/// Encode data as a DataUpdate (convenience wrapper)
///
/// This is equivalent to `to_bytes(data, true)`.
///
/// # Arguments
/// * `data` - Any serializable data
///
/// # Returns
/// UTF-8 bytes with Constellation prefix
pub fn encode_data_update<T: Serialize>(data: &T) -> Result<Vec<u8>> {
    to_bytes(data, true)
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn test_to_bytes_regular() {
        let data = json!({"id": "test", "value": 42});
        let bytes = to_bytes(&data, false).unwrap();
        let s = String::from_utf8(bytes).unwrap();
        assert_eq!(s, r#"{"id":"test","value":42}"#);
    }

    #[test]
    fn test_to_bytes_data_update() {
        let data = json!({"id": "test"});
        let bytes = to_bytes(&data, true).unwrap();
        let s = String::from_utf8(bytes).unwrap();
        assert!(s.starts_with("\x19Constellation Signed Data:\n"));
        assert!(s.contains('\n'));
    }

    #[test]
    fn test_encode_data_update() {
        let data = json!({"id": "test"});
        let bytes = encode_data_update(&data).unwrap();
        let s = String::from_utf8(bytes).unwrap();
        assert!(s.starts_with("\x19Constellation Signed Data:\n"));
    }
}

File: packages/typescript-core/src/binary.ts
/**
 * Binary Encoding
 *
 * Converts JSON data to binary format for cryptographic operations.
 * Supports both regular encoding and DataUpdate encoding with Constellation prefix.
 */

import { canonicalize } from './canonicalize';
import { CONSTELLATION_PREFIX } from './types';
import type { SigningMode } from './types';

/**
 * Convert data to binary bytes for signing
 *
 * For regular data:
 *   JSON -> RFC 8785 canonicalization -> UTF-8 bytes
 *
 * For DataUpdate (isDataUpdate=true or mode='dataUpdate'):
 *   JSON -> RFC 8785 -> UTF-8 -> Base64 -> prepend Constellation prefix -> UTF-8 bytes
 *
 * @param data - Any JSON-serializable object
 * @param mode - SigningMode ('standard' | 'dataUpdate') or boolean for backward compat
 * @returns Binary bytes as Uint8Array
 *
 * @example
 * ```typescript
 * // Regular encoding
 * const bytes = toBytes({ action: 'test' });
 *
 * // DataUpdate encoding (new API)
 * const updateBytes = toBytes({ action: 'test' }, 'dataUpdate');
 *
 * // DataUpdate encoding (legacy API)
 * const updateBytes2 = toBytes({ action: 'test' }, true);
 * ```
 */
export function toBytes<T>(data: T, mode: SigningMode | boolean = false): Uint8Array {
  const isDataUpdate = typeof mode === 'string' ? mode === 'dataUpdate' : mode;
  const canonicalJson = canonicalize(data);
  const utf8Bytes = new TextEncoder().encode(canonicalJson);

  if (isDataUpdate) {
    // Base64 encode the UTF-8 bytes
    const base64String = Buffer.from(utf8Bytes).toString('base64');
    // Create the wrapped string with Constellation prefix
    const wrappedString = `${CONSTELLATION_PREFIX}${base64String.length}\n${base64String}`;
    return new TextEncoder().encode(wrappedString);
  }

  return utf8Bytes;
}

/**
 * Encode data as a DataUpdate with Constellation prefix
 *
 * This is equivalent to `toBytes(data, true)`.
 *
 * @param data - Any JSON-serializable object
 * @returns Binary bytes with Constellation prefix
 */
export function encodeDataUpdate<T>(data: T): Uint8Array {
  return toBytes(data, true);
}
